### DISCUSSION TITLE
Help debugging crash loop on node

### DISCUSSION THREAD

**Discussion by bdkuhman** [longhorn/longhorn#6904] [Q&A]
Alright: I acknowledge this is a little strange, and definitely not a standard install, but I'm starting to run out of ideas and I'm starting to run into issues that don't seem to have a clear cause, and logging isn't very helpful.

This node is deployed as a docker container in Docker Desktop on Windows.

This is all still a WIP/POC so I don't have everything documented/setup yet; but I have longhorn installed on other nodes, and they're working just fine.

I need some help figuring out exactly why longhorn keeps restarting.

### Versions:
**Longhorn**: Swapped to 1.4.4-rc.1 from 1.5.1 in the hopes this is/was related to #6829  
**Kernel**: [wsl-6.1.21.2](https://github.com/microsoft/WSL2-Linux-Kernel/tree/linux-msft-wsl-6.1.21.2) custom compiled with **xt_recent** and **iscsi_tcp** kernel modules enabled; along with [these patches](https://lore.kernel.org/all/20230506232930.195451-4-cleech@redhat.com/T/) to allow iscsi to be network_namespace aware  
**WSL**  2.0.4.0 (pre-release to fix cgroups bug)  
**Kubelet**: v1.27.6+k3s1  with wireguard-native flannel backend
**Docker**: 24.0.5, build 24.0.5-0ubuntu1~20.04.1  
Docker image: Modified from k3d to install nvidia drivers/toolkit , along with installing iscsi / starting with systemd.   

In 1.5.1, dmesg would have some messages about the device getting detached and remounted read-only; which is why i swapped to 1.4.4; and I don't see the mount/unmount messages anymore, But the containers still crash/restart constantly.

Basically what's happening is everything will start up fine, but then failures will cascade through the pods; once they're all running, the node reports as ready and as replicas get pulled into the node, volume starts to be attached, then detached and then /something/ suddenly crashes and basically all the logs will say are "Receive terminated to exit", and it seems that the manager starts to force-restart everything, but nothing ever fully starts up again.

everything in the logs seem as normal as possible; Nothing obvious in dmesg, journalctl; as far as i can tell. To me it looks like something's sending a kill/restart signal, but I'm not sure what, or why.

**Comment by bdkuhman** [comment]
Collected logs:
[csi_plugin.log](https://github.com/longhorn/longhorn/files/12922244/csi_plugin.log)
[manager.log](https://github.com/longhorn/longhorn/files/12922245/manager.log)
[engineimage.log](https://github.com/longhorn/longhorn/files/12922246/engineimage.log)
[instance-manager-e.log](https://github.com/longhorn/longhorn/files/12922254/instance-manager-e.log)
[instance-manager-r.log](https://github.com/longhorn/longhorn/files/12922255/instance-manager-r.log)

**Reply by bdkuhman** [reply]
I may have figured /something/ out; 
I had an old 1.5.1 instance manager still running;
After deleting it, and the new 1.4.4 manager spawning, I get it to mount with periodic I/O errors in dmesg now... so back to where we were!

**Comment by DamiaSan** [comment]
Can you post the logs after deleting old 1.5.1 instance manager?

**Comment by c3y1huang** [comment]
Longhorn does not support downgrading. Despite this, Longhorn v1.5.0 added upgrade path enforcement to prevent it. Not sure why it's not blocking downgrading to 1.4.4-rc.1.

cc @mantissahz

**Reply by mantissahz** [reply]
Yes, we have upgrade path enforcement from v1.5.0.
The blocking logic starts from v1.5.0, so it is not contained in the v1.4.x longhorn-manager images and therefore v1.4.4-rc1 does not block the downgrading.

@bdkuhman there are some difference components created/deleted between v1.5.1 and v1.4.4-rc1.
From manager.log, there is no longhorn-engine:v1.5.1 for old PVCs and that caused instance-manager failed.

**Reply by bdkuhman** [reply]
I saw that longhorn doesn't support downgrading, and I was just going ahead with it to see if anything changed;
(and WSL for that matter, but I'm stubborn and here we are... 🙃)

With that said, other nodes in the cluster haven't had an issue other than not scheduling replicas w/ the aforementioned log messages, and their managers don't crash, but nevertheless I went ahead and nuked all the volumes, and reinstalled 1.4.4-rc1.

So far no real change, all other nodes are perfectly happy. longhorn on WSL will take a couple tries to startup, once it does; it can mount, but it looks like various components constantly are restarting/never get started and I don't see any obvious errors other than received signals. No i/o errors or anything that I've found.
Logs are very similar to what I posted earlier, minus the 1.5.1 engine errors.

Watching things, It looks like the engine image receives a signal, rm's the binary, and then it doesn't get restarted and then the other components don't have a bin and the whole cycle repeats, but that could just be one part of this and not what starts it.

I'll try to get more logs/maybe post some of this config up to github this weekend, but we'll see.

Things start to settle out here, but I never see the additional csi pods created. 
![image](https://github.com/longhorn/longhorn/assets/5903052/c5b98bcf-6b4b-4bfe-aefa-4560b6439e95)

**Reply by innobead** [reply]
@mantissahz please check this thread.

**Comment by bdkuhman** [comment]
I've published what's hopefully enough to get setup; 
https://github.com/bdkuhman/wsl2-longhorn

I'd love for anyone to take a whack at this.

somewhat in the process of updating everything, so versions might float around, but as of now I'm on wsl 2.0.6.0. 
Right now the instance managers don't spawn anymore, but we'll see.

Also it might be important, but the cluster is running with the wireguard-native flannel backend

**Reply by bdkuhman** [reply]
I think I figured it out finally.
The above config actually seems to work after some dns tweaking and a restart.

My original configuration involved adding nvidia things to get gpu support; (still in the process of setting up, will update the repo eventually)

but-- the changes involved adding a new default containerd runtime, of which the runtime options did not include 
` SystemdCgroup = true`
since I was pulling an example from somewhere else.

I'd still maybe like some tips on debugging these sort of things, or if there's anything for longhorn to check cgroups are enabled, I think it'd be handy to catch these kinds of problems.
